### PR TITLE
Add FLAC importing via dr_flac

### PR DESCRIPTION
**Pull request by Alicemargatroid**
Exactly as it says on the tin, very similar to StbVorbisImporter, provides FLAC loading via dr_flac.

dr_flac can also load FLAC that is stored in an OGG container, but I don't know how to provide that rather complex plugin dependency, so I've only added support if the file extension is ".flac".

While WAV provides the same use case of lossless audio, FLAC is considerably smaller and faster to decode, which could be important on mobile devices.

### CODE REVIEW

File: src/MagnumPlugins/AnyAudioImporter/AnyImporter.cpp [hunk 1]
```diff
@@ -53,6 +53,8 @@ void AnyImporter::doOpenFile(const std::string& filename) {
         plugin = "VorbisAudioImporter";
     else if(Utility::String::endsWith(filename, ".wav"))
         plugin = "WavAudioImporter";
+    else if(Utility::String::endsWith(filename, ".flac"))
+	plugin = "FlacAudioImporter";
```
Comment (mosra) [inline src/MagnumPlugins/AnyAudioImporter/AnyImporter.cpp]: :+1:

File: src/MagnumPlugins/DrFlacAudioImporter/DrFlacImporter.cpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+void DrFlacImporter::doOpenData(Containers::ArrayView<const char> data) {
+
+    drflac* handle = drflac_open_memory(reinterpret_cast<const UnsignedByte*>(data.data()), data.size());
+
+    if(handle == NULL) {
+        Error() << "Audio::DrFlacImporter::openData(): failed to open and decode FLAC data";
+        return;
+    }
+
+    _handle = (void*)handle;
```
Comment (mosra) [inline src/MagnumPlugins/DrFlacAudioImporter/DrFlacImporter.cpp]: If the opening fails on "unsupported channel count" below, the file is not considered being opened (`doIsOpened()` returns `false`) and thus `doClose()` would never get called, causing the `_handle` to leak.

I would save the handle only after everything is successful (i.e. along with saving the `_data` below) and not here. Then also call `drflac_close()` explicitly after the `Error()` call below.

File: src/MagnumPlugins/DrFlacAudioImporter/DrFlacImporter.h [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+This plugin is built if `WITH_DRFLACAUDIOIMPORTER` is enabled when building
+Magnum. To use dynamic plugin, you need to load `DrFlacAudioImporter` plugin
+from `MAGNUM_PLUGINS_AUDIOIMPORTER_DIR`. To use static plugin or use this as a
+dependency of another plugin, you need to request `DrFlacAudioImporter`
+component of `MagnumPlugins` package in CMake and link to
+`MagnumPlugins::DrFlacAudioImporter`.
+
+This plugins provides `VorbisAudioImporter`, but note that this plugin doesn't
+have complete support for all format quirks and the performance might be worse
+than when using plugin dedicated for given format.
```
Comment (mosra) [inline src/MagnumPlugins/DrFlacAudioImporter/DrFlacImporter.h]: I think you forgot to update some parts here (still referencing OGG/vorbis/stb in some parts).

File: src/MagnumPlugins/DrFlacAudioImporter/DrFlacImporter.h [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        MAGNUM_DRFLACAUDIOIMPORTER_LOCAL void doClose() override;
+
+        MAGNUM_DRFLACAUDIOIMPORTER_LOCAL Buffer::Format doFormat() const override;
+        MAGNUM_DRFLACAUDIOIMPORTER_LOCAL UnsignedInt doFrequency() const override;
+        MAGNUM_DRFLACAUDIOIMPORTER_LOCAL Containers::Array<char> doData() override;
+
+        Containers::Array<char> _data;
+        Buffer::Format _format;
+        UnsignedInt _frequency;
+	void* _handle;
```
Comment (mosra) [inline src/MagnumPlugins/DrFlacAudioImporter/DrFlacImporter.h]: Could you say

```cpp
struct drflac* _handle;
```

here? It's just a pointer so a forward declaration is enough, and then you also avoid those ugly `void*` casts in the implementation. Also the indentation is a tiny bit off ;)

**Comment by mosra** [comment]
Oh cool! This is great :)

Regarding test coverage -- would it be possible for you to create also some minimal test case (just some minimal one-second file, only verifying that the third party code is able to execute on all platforms without crashing, nothing fancy) and then enable the plugin in all `package/ci/travis-*.sh` and `package/ci/appveyor-*.bat` CI configurations?

I would be very happy to merge it afterwards.

File: src/MagnumPlugins/AnyAudioImporter/AnyImporter.cpp (hunk 1 above)
Comment (LB--) [inline src/MagnumPlugins/AnyAudioImporter/AnyImporter.cpp]: Looks like the indentation is mismatched

Comment (Alicemargatroid) [inline src/MagnumPlugins/AnyAudioImporter/AnyImporter.cpp]: @LB-- Looks like that happened in a few locations, I'll patch it shortly. My nano instance isn't properly configured yet, and my eyes aren't what they used to be!

File: src/MagnumPlugins/DrFlacAudioImporter/DrFlacImporter.cpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+    int32_t* decodedData = handle->pDecodedSamples;
+    uint64_t samples = handle->totalSampleCount;
+    uint32_t frequency = handle->sampleRate;
+    uint8_t numChannels = handle->channels;
+    uint8_t bitsPerSample = handle->bitsPerSample;
+
+    _frequency = frequency;
+
+    Containers::Array<char> tempData{reinterpret_cast<char*>(decodedData), size_t(samples*numChannels*2)};
```
Comment (Alicemargatroid) [inline src/MagnumPlugins/DrFlacAudioImporter/DrFlacImporter.cpp]: @mosra Question I wasn't easily able to resolve myself; does `Containers::Array<char>` copy `decodedData`, or is there a way to make it do so? Then we could get rid of the entire `_handle`, and I wouldn't have to manage it at all.

Comment (mosra) [inline src/MagnumPlugins/DrFlacAudioImporter/DrFlacImporter.cpp]: Yeah, that would be even better! `Containers::Array` doesn't copy the data in any way (design decision), so you need to use `std::copy()` or something like in the `doData()` below.

Comment (Alicemargatroid) [inline src/MagnumPlugins/DrFlacAudioImporter/DrFlacImporter.cpp]: Just as I submit different changes! Let me rewrite it that way.

**Comment by Alicemargatroid** [comment]
@mosra I think the last push should resolve all issues besides test cases.

I will attempt to write some small test cases as soon as possible, hopefully by the end of the day.

**Comment by Alicemargatroid** [comment]
@mosra Three tests added, test FLAC's are licensed CC0/public domain.

How do I enable the tests in our CI? Think that is the last step.

**Comment by mosra** [comment]
Like this:

```patch
diff --git a/package/ci/appveyor-desktop-mingw.bat b/package/ci/appveyor-desktop-mingw.bat
index 473377a..3f7212a 100644
--- a/package/ci/appveyor-desktop-mingw.bat
+++ b/package/ci/appveyor-desktop-mingw.bat
@@ -67,6 +67,7 @@ cmake .. ^
     -DWITH_ANYIMAGEIMPORTER=ON ^
     -DWITH_ANYSCENEIMPORTER=ON ^
     -DWITH_DDSIMPORTER=ON ^
+    -DWITH_DRFLACAUDIOIMPORTER=ON ^
     -DWITH_FREETYPEFONT=OFF ^
     -DWITH_HARFBUZZFONT=OFF ^
     -DWITH_JPEGIMPORTER=ON ^
diff --git a/package/ci/appveyor-desktop.bat b/package/ci/appveyor-desktop.bat
index fe71d50..24fa4a1 100644
--- a/package/ci/appveyor-desktop.bat
+++ b/package/ci/appveyor-desktop.bat
@@ -66,6 +66,7 @@ cmake .. ^
     -DWITH_ANYIMAGEIMPORTER=ON ^
     -DWITH_ANYSCENEIMPORTER=ON ^
     -DWITH_DDSIMPORTER=ON ^
+    -DWITH_DRFLACAUDIOIMPORTER=ON ^
     -DWITH_FREETYPEFONT=OFF ^
     -DWITH_HARFBUZZFONT=OFF ^
     -DWITH_JPEGIMPORTER=ON ^
diff --git a/package/ci/appveyor-rt.bat b/package/ci/appveyor-rt.bat
index 4dfc306..ea3b44a 100644
--- a/package/ci/appveyor-rt.bat
+++ b/package/ci/appveyor-rt.bat
@@ -78,6 +78,7 @@ cmake .. ^
     -DWITH_ANYIMAGEIMPORTER=ON ^
     -DWITH_ANYSCENEIMPORTER=ON ^
     -DWITH_DDSIMPORTER=ON ^
+    -DWITH_DRFLACAUDIOIMPORTER=OFF ^
     -DWITH_FREETYPEFONT=OFF ^
     -DWITH_HARFBUZZFONT=OFF ^
     -DWITH_JPEGIMPORTER=OFF ^
diff --git a/package/ci/travis-desktop.sh b/package/ci/travis-desktop.sh
index 06cc278..405203c 100755
--- a/package/ci/travis-desktop.sh
+++ b/package/ci/travis-desktop.sh
@@ -50,6 +50,7 @@ cmake .. \
     -DWITH_ANYSCENEIMPORTER=ON \
     -DWITH_COLLADAIMPORTER=$WITH_COLLADAIMPORTER \
     -DWITH_DDSIMPORTER=ON \
+    -DWITH_DRFLACAUDIOIMPORTER=ON \
     -DWITH_FREETYPEFONT=ON \
     -DWITH_HARFBUZZFONT=ON \
     -DWITH_JPEGIMPORTER=ON \
diff --git a/package/ci/travis-emscripten.sh b/package/ci/travis-emscripten.sh
index 014a7ec..c85967c 100755
--- a/package/ci/travis-emscripten.sh
+++ b/package/ci/travis-emscripten.sh
@@ -75,6 +75,7 @@ cmake .. \
     -DWITH_ANYSCENEIMPORTER=ON \
     -DWITH_COLLADAIMPORTER=OFF \
     -DWITH_DDSIMPORTER=ON \
+    -DWITH_DRFLACAUDIOIMPORTER=ON \
     -DWITH_FREETYPEFONT=OFF \
     -DWITH_HARFBUZZFONT=OFF \
     -DWITH_JPEGIMPORTER=OFF \
diff --git a/package/ci/travis-ios-simulator.sh b/package/ci/travis-ios-simulator.sh
index 6694827..d12647b 100755
--- a/package/ci/travis-ios-simulator.sh
+++ b/package/ci/travis-ios-simulator.sh
@@ -78,6 +78,7 @@ cmake .. \
     -DWITH_ANYSCENEIMPORTER=ON \
     -DWITH_COLLADAIMPORTER=OFF \
     -DWITH_DDSIMPORTER=ON \
+    -DWITH_DRFLACAUDIOIMPORTER=ON \
     -DWITH_FREETYPEFONT=OFF \
     -DWITH_HARFBUZZFONT=OFF \
     -DWITH_JPEGIMPORTER=OFF \

```

**Comment by mosra** [comment]
Squased into one commit and merged into master in 87d81603f3d7907772c9db29ad3270464f0a4b0d. I did some additional boilerplate updates so it is now also available through `FindMagnumPlugins.cmake`.

Thanks a lot for this :)

**Pull request closed without merging**